kernel/src/hil/gpio.rs

```diff
@@ -1,67 +1,289 @@
-//! Interface for direct control of GPIO pins.
+use crate::common::cells::OptionalCell;
+use crate::ReturnCode;
+
+use core::cell::Cell;
 
 /// Enum for configuring any pull-up or pull-down resistors on the GPIO pin.
 #[derive(Debug)]
-pub enum InputMode {
+pub enum FloatingState {
     PullUp,
     PullDown,
     PullNone,
 }
 
 /// Enum for selecting which edge to trigger interrupts on.
 #[derive(Debug)]
-pub enum InterruptMode {
+pub enum InterruptEdge {
     RisingEdge,
     FallingEdge,
     EitherEdge,
 }
 
-pub trait PinCtl {
-    /// Configure whether the pin should have a pull-up or pull-down resistor or
-    /// neither.
-    fn set_input_mode(&self, _: InputMode);
+/// Enum for which state the pin is in. Some MCUs can support Input/Output pins,
+/// so this is a valid option. `Function` means the pin has been configured to
+/// a special function. Determining which function it outside the scope of the HIL,
+/// and should instead use a chip-specific API.
+#[derive(Debug)]
+pub enum Configuration {
+    LowPower,
+    Input,
+    Output,
+    InputOutput,
+    Function, // Chip-specific, requires chip-specific API for more detail
+    Unknown,
 }
 
-/// Interface for synchronous GPIO pins.
-pub trait Pin {
-    /// Configure the GPIO pin as an output pin.
-    fn make_output(&self);
+/// The Pin trait allows a pin to be used as either input
+/// or output and to be configured.
+pub trait Pin: Input + Output + Configure {}
+
+/// The InterruptPin trait allows a pin to be used as either
+/// input or output and also to source interrupts.
+pub trait InterruptPin: Pin + Interrupt {}
+
+/// The InterruptValuePin trait allows a pin to be used as
+/// either input or output and also to source interrupts which
+/// pass a value.
+pub trait InterruptValuePin: Pin + InterruptWithValue {}
 
-    /// Configure the GPIO pin as an input pin.
-    fn make_input(&self);
+pub trait Configure {
```

Comment:
trait documentation please,

Also, I find the `Configuration trait` hard to understand with methods such as `make_output` in addition being a super trait to `Pin`

Reply:
Configure allows you to change how the pin operates; it's broken out from Input and Output so that you can give access to read or write a pin without also granting the ability to change its operating mode. The Pin trait is more full access to a Pin: you can configure it, read it, and write it.

Reply:
It's the old `PinCtl`. I think it makes a lot of sense as a trait, I'm a bit confused by some of the specific methods or their signatures in it (we can discuss during the call?).

@niklasad1 totally agree though, the final version (before merging) should (and will) definitely have much more documentation.